### PR TITLE
 drop-in type check for rb_define_method

### PR DESCRIPTION
**Pull request by shyouhei**
The `rb_define_method` function takes a pointer to `ANYARGS`-ed functions, which in fact varies 18 different prototypes.  We still need to preserve `ANYARGS` for storages but why not check the consistencies if possible.

### CODE REVIEW

**Comment by shyouhei** [comment]
Will merge this.

File: eval.c [hunk 1]
```diff
@@ -1938,16 +1944,34 @@ f_current_dirname(void)
...
     rb_define_virtual_variable("$@", errat_getter, errat_setter);
     rb_define_virtual_variable("$!", errinfo_getter, 0);
 
-    rb_define_global_function("raise", rb_f_raise, -1);
-    rb_define_global_function("fail", rb_f_raise, -1);
+    rb_define_global_function("raise", f_raise, -1);
+    rb_define_global_function("fail", f_raise, -1);
 
-    rb_define_global_function("global_variables", rb_f_global_variables, 0);	/* in variable.c */
+    rb_define_global_function("global_variables", f_global_variables, 0);
```
Comment (nobu) [inline eval.c]: Probably, this may prevent `global_variables` being documented?

File: include/ruby/intern.h [hunk 2]
```diff
@@ -421,7 +421,7 @@ int rb_method_basic_definition_p(VALUE, ID);
 VALUE rb_eval_cmd(VALUE, VALUE, int);
 int rb_obj_respond_to(VALUE, ID, int);
 int rb_respond_to(VALUE, ID);
-NORETURN(VALUE rb_f_notimplement(int argc, const VALUE *argv, VALUE obj));
+NORETURN(VALUE rb_f_notimplement(int argc, const VALUE *argv, VALUE obj, VALUE marker));
 #if !defined(RUBY_EXPORT) && defined(_WIN32)
 RUBY_EXTERN VALUE (*const rb_f_notimplement_)(int, const VALUE *, VALUE);
```
Comment (nobu) [inline include/ruby/intern.h]: This declaration may need `marker` too.

Comment (shyouhei) [reply include/ruby/intern.h]: Good catch, yes.

File: eval.c (hunk 1 above)
Comment (shyouhei) [reply eval.c]: moved documents @ 9a242d76a76f8cc55d3f0d4553d5c27eb6ffc6b8

**Comment by k0kubun** [comment]
Seems like merged in https://github.com/ruby/ruby/commit/e4be2fda3dbbfdb1f2ace697c96cf6bdd7dfef21 and so on.

By the way, for the master branch, you can use "Rebase and merge" button in a pull request page  to merge commits like your way of merging, as we have https://bugs.ruby-lang.org/issues/16094 experimentally.

**Pull request closed without merging**